alts/src/main/java/io/grpc/alts/internal/TsiHandshakeHandler.java

```diff
@@ -137,13 +146,37 @@ protected void decode(ChannelHandlerContext ctx, ByteBuf in, List<Object> out) t
   }
 
   @Override
-  public void userEventTriggered(ChannelHandlerContext ctx, Object evt) throws Exception {
+  public void userEventTriggered(final ChannelHandlerContext ctx, Object evt) throws Exception {
     if (evt instanceof ProtocolNegotiationEvent) {
       checkState(pne == null, "negotiation already started");
       pne = (ProtocolNegotiationEvent) evt;
       InternalProtocolNegotiators.negotiationLogger(ctx)
           .log(ChannelLogLevel.INFO, "TsiHandshake started");
-      sendHandshake(ctx);
+      ChannelFuture acquire = semaphore.acquire(ctx);
+      if (acquire.isSuccess()) {
+        semaphoreAcquired = true;
+        sendHandshake(ctx);
+      } else {
+        acquire.addListener(new ChannelFutureListener() {
+          @Override public void operationComplete(ChannelFuture future) {
+            if (!future.isSuccess()) {
+              ctx.fireExceptionCaught(future.cause());
+              return;
+            }
+            if (ctx.isRemoved()) {
+              semaphore.release();
+              return;
+            }
+            semaphoreAcquired = true;
+            try {
+              sendHandshake(ctx);
+            } catch (Exception ex) {
+              ctx.fireExceptionCaught(ex);
+            }
+            ctx.flush();
```

Comment:
Just curious, why we want to flush here?

Reply:
Without the flush no data would be sent. Probably ever. Normally the flush is handled at a higher level: RPC senders will do the flush themselves, and we do a flush after each batch of reads (in case the reads triggered writes). Since the event triggering this code is not on the normal read and write paths, there wouldn't be a flush naturally occurring.